src/daft-functions-utf8/src/count_matches.rs

```diff
@@ -96,6 +99,75 @@ impl ScalarUDF for CountMatches {
     }
 }
 
+#[typetag::serde]
+impl ScalarUDF for CountMatchesRegex {
+    fn name(&self) -> &'static str {
+        "count_matches_regex"
+    }
+    fn call(&self, inputs: daft_dsl::functions::FunctionArgs<Series>) -> DaftResult<Series> {
+        let input = inputs.required((0, "input"))?;
+        let patterns = inputs.required((1, "patterns"))?;
+
+        let whole_words = inputs.optional("whole_words")?.map(|s| {
+            ensure!(s.data_type().is_boolean() && s.len() == 1, ValueError: "expected boolean literal for 'whole_words'");
+            Ok(s.bool().unwrap().get(0).unwrap())
+        }).transpose()?.unwrap_or(WHOLE_WORDS_DEFAULT_VALUE);
+        let case_sensitive = inputs.optional("case_sensitive")?.map(|s| {
+            ensure!(s.data_type().is_boolean() && s.len() == 1, ValueError: "expected boolean literal for 'case_sensitive'");
+            Ok(s.bool().unwrap().get(0).unwrap())
+        }).transpose()?.unwrap_or(CASE_SENSITIVE_DEFAULT_VALUE);
+
+        ensure!(patterns.len() == 1, ValueError: "Cannot set `patterns` in `count_matches_regex` to an Expression. Only string is currently supported.");
+
+        input.with_utf8_array(|arr| match patterns.data_type() {
+            DataType::Utf8 => patterns.with_utf8_array(|pattern_arr| {
+                Ok(
+                    regex_count_matches_impl(arr, pattern_arr, whole_words, case_sensitive)?
+                        .into_series(),
+                )
+            }),
+            patterns_dtype => Err(DaftError::ValueError(format!(
+                "expected string for 'patterns', got {}",
+                patterns_dtype
+            ))),
+        })
+    }
+
+    fn get_return_field(
+        &self,
+        inputs: FunctionArgs<ExprRef>,
+        schema: &Schema,
+    ) -> DaftResult<Field> {
+        let input = inputs.required((0, "input"))?.to_field(schema)?;
+        let patterns = inputs.required((1, "patterns"))?.to_field(schema)?;
+        ensure!(patterns.dtype.is_string(), ValueError: "expected string for 'patterns', got {}", patterns.dtype);
+
+        if let Some(whole_words) = inputs.optional("whole_words")? {
+            whole_words
+                .as_literal()
+                .and_then(|lit| lit.as_bool())
+                .ok_or(DaftError::ValueError(
+                    "expected boolean literal for 'whole_words'".to_string(),
+                ))?;
+        }
+
+        if let Some(case_sensitive) = inputs.optional("case_sensitive")? {
+            case_sensitive
+                .as_literal()
+                .and_then(|lit| lit.as_bool())
+                .ok_or(DaftError::ValueError(
+                    "expected boolean literal for 'case_sensitive'".to_string(),
+                ))?;
+        }
```

Comment:
Now that we've split the function out of the non-regex `count_matches`, could we remove these parameters entirely?

Reply:
Yeah noticed that a bit too late